### PR TITLE
BUG: Fix scope get to use hashmap lookup instead of list lookup

### PR DESCRIPTION
**Pull request by LeeTZ**
Overview
------
This PR aims to fix a bug in looking up an `op` in `scope`.  The current implementation is using a list lookup to look up items in `iter(_items)`. This is slower than a hashmap lookup, also it may cause bugs for objects if their equality is ill-defined.

Example
-------
```
a = ibis.interval(days=1).op()
b = ibis.interval(hours=1).op()
>>> a
Literal(1, Interval*value_type=int8, unit='D', nullable=True)
>>> a == b
True
```
In this case, `Literal` is ill-defined for equality. And if we have only `a` in scope`. `b in iter(scope._items)` will be `True` while `b in scope._items` return false. 

Tests
-------
This is tested in `ibis/pandas/tests/tes_core.py` with a new test case `test_scope_look_up`.

### CODE REVIEW

File: ibis/expr/scope.py [hunk 1]
```diff
@@ -122,10 +122,10 @@ def merge_scope(self, other_scope: 'Scope', overwrite=False) -> 'Scope':
         """
         result = Scope()
 
-        for op in self.items():
+        for op in self._items:
```
Comment (jreback) [inline ibis/expr/scope.py:125]: should be `_items.keys()` no?

File: ibis/expr/scope.py [hunk 2]
```diff
@@ -122,10 +122,10 @@ def merge_scope(self, other_scope: 'Scope', overwrite=False) -> 'Scope':
...
             result[op] = self[op]
 
-        for op in other_scope.items():
+        for op in other_scope._items:
```
Comment (jreback) [inline ibis/expr/scope.py:128]: same

File: ibis/expr/scope.py [hunk 3]
```diff
@@ -153,7 +153,7 @@ def merge_scopes(
             a new Scope instance with items in two scope merged.
         """
         result = Scope()
-        for op in self.items():
+        for op in self._items:
```
Comment (jreback) [inline ibis/expr/scope.py:156]: same

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by icexelloss** [comment]
I found this class to be a little confusing. It looks like we are emulating dict interface by implementing `__setitem__` and `__getitem__` in addition to `get_value` and `merge_scope`, but:

* It didn't implement other dict interfaces, such as `__contains__`
* Scope.items() returns the keys in the scope (the ops), but dict.items() returns the dict entries (k, v pair), why do we make them different if we are emulating dict interface?
* It's not obviously clear to me if this class is defined to be mutable or immutable - on the one hand, `merge_scope` seems to not mutating exising scope, but `__setitem__` indicates this class is mutable.

I think clarifying these questions would be helpful to make it more clear how to use this class.

File: ibis/expr/scope.py (hunk 1 above)
Comment (LeeTZ) [reply ibis/expr/scope.py:125]: I am not sure why using `.keys()` is better here, `_items` is a dict and `op` is its key. Isn't a simple `in` enough?

Comment (jreback) [reply ibis/expr/scope.py:125]: because its not obvious that you _items is actually a dict from just looknig here, whereas .keys() is very explict

**Comment by LeeTZ** [comment]
Thanks, @icexelloss ,
This class Scope is refactored from a dict. So it makes sense to implement some interfaces at the first glance. However I have to point out that from the perspective of a user of this class, the key is important, but the value is not. Since value is a data structure that we made up to wrap time context and value, we do need API to get the real value of the cached result, but not necessarily the API to get `{"value":v,"timecontext":t}`.  
Also, for `set`, most scope usages for setting in Ibis is merging scopes generated by different functions. This class tried to preserve this behavior so `merge` is really the `set` method for this class.

So given this `get` and `set` use case, I will say without `__get_item__` , `__set_item__` or any dict like interface, this class works. Dict interfaces are helpful in some implementation of the internal method inside this class, but it does add the ambiguity of how to use this class.

For @icexelloss's question about `__contain__` and `items()`, would you mind sharing your thoughts?@jreback

File: ibis/expr/scope.py (hunk 1 above)
Comment (LeeTZ) [reply ibis/expr/scope.py]: OK, fixed it.

**Comment by LeeTZ** [comment]
I refactored a little bit.
  - removed `__get_item__`. As I commented above, we don't care about the object associated with `op`, it is an implementation detail for this class and we should hide it from users. We exposed `get_value` to get the cached result for a given `op`.
  - removed `__set_item__`. We want to make this class immutable and always use `merge` for adding new items into `Scope`. This is the usage case in Ibis everywhere (by calling `toolz.merge`, before time context refactor )
  - add `__contains__` One thing we need is to test whether an `op` is in Scope. This replaces `op in scope._items.keys()` or `scope.get_value(op)`
  - I think this class doesn't fully emulate a dict interface since we have made it immutable. thus not all dict interfaces are implemented.


To make this class clearer, these are the exposed API for `Scope` class:
1. `new` a 'Scope`: user should use `Scope()` to create an empty `Scope`, or call `make_scope` with a pair of `op, result, time context` to create a new Scope with one item in it.
2. `get` in `Scope`:  user needs to call `get_value` to get the cached result for a given `op`.
3. `set` in `Scope`:  user always uses `merge_scope` or `merge_scopes` to set values.

File: ibis/expr/scope.py [hunk 4]
```diff
@@ -48,14 +48,8 @@ class Scope:
...
-    def items(self):
-        return iter(self._items)
-
-    def __getitem__(self, op: Node) -> ScopeItem:
-        return self._items[op]
-
-    def __setitem__(self, op: Node, value: Any) -> None:
-        self._items[op] = value
+    def __contains__(self, op):
+        return True if op in self._items.keys() else False
```
Comment (icexelloss) [inline ibis/expr/scope.py:52]:
```
def __contains__(self, op):
    return op in self.items
```
?

File: ibis/expr/scope.py [hunk 5]
```diff
@@ -122,17 +116,17 @@ def merge_scope(self, other_scope: 'Scope', overwrite=False) -> 'Scope':
         """
         result = Scope()
 
-        for op in self.items():
-            result[op] = self[op]
+        for op in self._items.keys():
```
Comment (icexelloss) [inline ibis/expr/scope.py:119]: Is this basically what you want?
https://stackoverflow.com/questions/6354436/python-dictionary-merge-by-updating-but-not-overwriting-if-value-exists

```
new_items = {}

if overwrite:
    new_items = dict(list(self._items.items()) + list(other._items.items()))
else:
    new_items = dict(list(other._items.items()) + list(self._items.items()))

return new Scope(new_items)
```

**Review by icexelloss — CHANGES_REQUESTED** [review]
Left some comments

**Comment by icexelloss** [comment]
I think it makes sense. Emulating dict interface is nice but no necessarily IMO.

File: ibis/expr/scope.py (hunk 5 above)
Comment (LeeTZ) [reply ibis/expr/scope.py:119]: This is nice but to update we cannot simply do these. We should think about time context and the logic is covered in `result.get_value(op, v.timecontext) is None`

File: ibis/expr/scope.py (hunk 4 above)
Comment (LeeTZ) [reply ibis/expr/scope.py:52]: Yes

File: ibis/expr/scope.py [hunk 6]
```diff
@@ -48,14 +48,8 @@ class Scope:
...
-    def items(self):
-        return iter(self._items)
-
-    def __getitem__(self, op: Node) -> ScopeItem:
-        return self._items[op]
-
-    def __setitem__(self, op: Node, value: Any) -> None:
-        self._items[op] = value
+    def __contains__(self, op):
+        return op in self._items.keys()
```
Comment (icexelloss) [inline ibis/expr/scope.py:52]: why not use 

```
return op in self._items?
```

File: ibis/expr/scope.py (hunk 5 above)
Comment (icexelloss) [reply ibis/expr/scope.py:119]: I see. Ok what you have is fine.

File: ibis/expr/scope.py (hunk 6 above)
Comment (icexelloss) [reply ibis/expr/scope.py:52]: Oh I see Jeff's comments. Ok this is fine.

**Review by icexelloss — APPROVED** [review]
LGTM

File: ibis/expr/scope.py (hunk 6 above)
Comment (jreback) [reply ibis/expr/scope.py:52]: hmm, no i agree this should be in `self._items`, only *iteration* should use the keys

File: ibis/expr/scope.py (hunk 5 above)
Comment (jreback) [reply ibis/expr/scope.py:119]: why don't we define
```
def __iter__(self):
    return iter(self._items.keys())
```

then 
```
for op in self:
   ...
````

will just work

or is this more confusing?

**Review by jreback — CHANGES_REQUESTED** [review]

File: ibis/expr/scope.py (hunk 5 above)
Comment (LeeTZ) [reply ibis/expr/scope.py:119]: This is fine, scope is indeed iterable, I will do that

File: ibis/expr/scope.py (hunk 6 above)
Comment (icexelloss) [reply ibis/expr/scope.py:52]: For my knowledge, is `k in self._items.keys` O(n) or O(1)? The type of self._items.keys is `dict_keys` so it's not clear to me if the `__contains__` method of `dict_key` is linear or constant.

Comment (LeeTZ) [reply ibis/expr/scope.py]: I looked around and seems that in Python 3.x, `k in self._items` and `k in self._items.keys()` are equivalent and they are all `O(1)`.

**Review by icexelloss — CHANGES_REQUESTED** [review]
Small comments

File: ibis/expr/scope.py (hunk 6 above)
Comment (LeeTZ) [reply ibis/expr/scope.py]: I tested with a Python 3.6 kernel
```
>>> dic = dict.fromkeys(range(10**5))
>>> %timeit 10000 in dic
35.8 ns ± 0.757 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
>>> %timeit 10000 in dic.keys()
74.7 ns ± 0.385 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```
seems `keys()` is still 2x slower than testing `in` for dict directly

File: ibis/expr/scope.py [hunk 7]
```diff
@@ -116,10 +119,10 @@ def merge_scope(self, other_scope: 'Scope', overwrite=False) -> 'Scope':
         """
         result = Scope()
 
-        for op in self._items.keys():
+        for op in self:
             result._items[op] = self._items[op]
 
-        for op in other_scope._items.keys():
+        for op in other_scope._items:
```
[the remark this answers is not in the document]
Comment (LeeTZ) [reply ibis/expr/scope.py]: It is, thanks

**Review by jreback — CHANGES_REQUESTED** [review]
can you add this PR number to the original scope update in the release notes

**Review by icexelloss — APPROVED** [review]
+1

**Review by jreback — APPROVED** [review]
lgtm. let's merge on green.

**Comment by LeeTZ** [comment]
Thank you all for reviewing! Green now @jreback .

**Pull request merged**